core/templates/dev/head/components/SearchBarDirective.js

```diff
@@ -151,11 +164,49 @@ oppia.directive('searchBar', [function() {
             }
 
             _updateSelectionDetails('languageCodes');
+
+            if (Boolean(urlService.getUrlParams().q)) {
+              _parseURL();
+            }
             _onSearchQueryChangeExec();
 
             _searchBarFullyLoaded = true;
           }
         );
+
+        var _parseURL = function() {
+          var urlSearch = decodeURI($window.location.search);
+          var query = urlSearch.replace('?q=', '');
+
+          // Grab language code(s) from URL, these are 2 lowercase letters.
+          var _languageCodeSelections = query.match(/"\w\w"/g);
```

Comment:
Probably a better logical flow would be: break the URL down into the query part, the language part, and the category part, in a mutually exclusive way. Then deal with each of these independently. Try not to operate on the entire query string each time; that can lead to errors.

It might be a good idea to write some karma tests for this -- feed in various URLs and make sure they are parsed correctly.